Major conventional overhaul

p4stoboy opened the pull request:
## Description

After reflecting on the PR #5 discussion and reviewing the DXBall codebase in more depth I feel that although it was in a better state to work on, it is still relatively unappealing to people being onboarded - some further low hanging adjustments have been identified which should make any proposed redesign or extension (some ideas in README) far more trivial.

#### General code adjustments
---
- Use const references for function parameters to avoid unnecessary copying and indicate intent regarding immutability.
- Use enum class instead of enum for better type safety and scope resolution.
- Use `std::array` instead of C-style arrays.
- Use `std::unique_ptr` or `std::shared_ptr` for dynamic memory management.
- Use `constexpr` for constants that can be evaluated at compile-time.
- Use static_cast instead of C-style casts for improved type safety.
‌
#### Structural changes
---
- Redesign header / declaration patterns to reduce dependency tree complexity and better reflect functional implementation.
- Move header files to `/include`.
- Redesign functional (.cpp) implementations to reflect new declaration silos.
‌
#### Conventional adjustments
---
- Implement doxygen style commentary over header declarations.
- Re-introduce (and redesign) procedural commentary to assist with navigating antipatterns remaining in codebase and assist with rapid onboarding (most of these comments are not conventional but are likely useful under the circumstances).

#### Other
---
- Update README



#### Type of change
Other (formatting, structuring & conventional adjustments)

#### How Has This Been Tested?
Built for linux and played through game.

#### Checklist
- [x] (?) My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (N/A)
- [x] My changes generate no new warnings
- [x] I have requested a review from peers via teams/trello on the Pull Request

Commit:
refactor(common): Adjust partitioning  / documentary comments / implement modern standards

redesign file structure to better reflect original implementation and make import pathing less interdependent, redesign and reimplement procedural comments, implement modern cpp standards where lacking (smart pointers, enum classes, static casting, constexpr) to enhance type safety and generally be more explicit.

Commit:
refactor(common): Adjust partitioning  / documentary comments / implement modern standards

redesign file structure to better reflect original implementation and make import pathing less interdependent, redesign and reimplement procedural comments, implement modern cpp standards where lacking (smart pointers, enum classes, static casting, constexpr) to enhance type safety and generally be more explicit.

Commit:
refactor(common): Add to README / reimplement TODO comment

Commit:
refactor(common): Add to README / reimplement TODO comment

Commit:
refactor(common): Update README

MHLoppy reviewed (commented):
Following up on my comment (on Teams) from my initial lookover earlier, this is actually more changes than I expected. A few of the more advanced changes I'm not sure any of the other students -- including me without an SE background -- can properly review so I'll leave that to Tien and your own more experienced judgement!

Of the things I feel I can evaluate, the readability improvements are great, especially the updated naming to be much more self-descriptive. I'm slightly surprised at how many non-header comments you've added (or added back) and I'm feeling a bit of whiplash now, but per previous discussion I'm not opposed to the current level of commenting for this project.

**Questions:**
* Can you give me a rundown of this usage of `constexpr` compared to `const` in this context? I did some skim-reading about it, but it's not something I've encountered before. In this usage (where it's isolated to simple variables in `global_constants.h`), is this basically just potentially free-performance due to possible compiler optimization?
* Is it worth git-excluding `scores.json` ? I assume every time anybody works on this game they'll likely end up populating at least one entry while testing, so wiping the current entries and then git-excluding it might make future PRs a bit neater by not having an extra file changed every time.

**Suggestions**
* A few times `json_read_<thing>` is called as part of the parameter of another function call. In cases where this makes the line of code quite long (e.g. line 21/22 of `high_scores.cpp`), it might be better to throw that result in a variable first and use that in the parameter instead, or just break up the line even without a variable. I'll leave it to you whether you think it's worthwhile to do, as I know that there are other sources of long line length still in there regardless.
* The comment on line 59 of `types.h` (not the line 60 one) is off by one space. I'm not going to block the PR for that but if you can be bothered, it'd be nice to fix it for neatness' sake.
* While you're still familiar with the codebase, I think it'd be worthwhile fleshing out your note in the readme about potential future improvements if you had anything else (at an overview-level of detail) to add. If what you said there is all of it already then no worries though.

---

Once you've made any final changes as you see fit (including _none_ if you want!) I'll approve it and bump you over to Tien.

p4stoboy commented:
Thanks Mark,

Re - the whiplash around the comments - that's totally fair; again - I'm not of the view that they're conventional and were this a project with a consistent team around it I would say they should be removed immediately, but after reflecting on your comments and the high degree of turnover in terms of who is working on these projects (combined with wildly varied levels of experience wrt SWE in general) I think they're probably a net benefit in this instance.

- `constexpr` just signals that the immutable value is known at compile time (ie. we can still assign `const int x = y * z` but if the values of `y` and `z` are derived from some non-deterministic run-time pattern then we can't know what the value of `x` is until it is assigned during execution). The same can be applied to functions etc where every possible set of arguments can be evaluated at compile time etc. And yes, fundamentally it's just a pattern to mitigate the levels of abstraction the compiler has to deal with and can enable more aggressive optimisations, minimise build size etc (ie. if `x` was instead a constexpr derived from values known at compile time then the compiler is not having to generate patterns to derive the potential values it could take).

- I like keeping scores in because it demonstrates that someone has played through the game since they pulled the repo (or it should), but if you created a ticket to ignore it I wouldn't object ;)

I'll take a look at the other stuff, the readme notes are likely finished but I definitely would like to fix that comment offset :)

Commit:
refactor(common): Adjustments based on PR feedback

- removed json_read_x() calls from inline arguments and assigned to named values before providing as function argument.
- fixed comment offsets in types.h

MHLoppy reviewed (approved):
I'm ambivalent about the scores so no worries keeping that in.

Approved! (though give a yell if you end up making any further code changes and I can give that a glance before Tien)
